go-controller/pkg/node/gateway_localnet.go

```diff
@@ -107,55 +123,50 @@ func (n *OvnNode) initLocalnetGateway(subnet *net.IPNet, nodeAnnotator kube.Anno
 		ChassisID:      chassisID,
 		InterfaceID:    ifaceID,
 		MACAddress:     macAddress,
-		IPAddresses:    []*net.IPNet{gatewayIPCIDR},
-		NextHops:       []net.IP{gatewayNextHop},
+		IPAddresses:    gatewayIfAddrs,
+		NextHops:       nextHops,
 		NodePortEnable: config.Gateway.NodeportEnable,
 	})
 	if err != nil {
 		return err
 	}
 
-	if utilnet.IsIPv6CIDR(subnet) {
-		// TODO - IPv6 hack ... for some reason neighbor discovery isn't working here, so hard code a
-		// MAC binding for the gateway IP address for now - need to debug this further
-		err = util.LinkNeighAdd(link, gatewayIP, macAddress)
-		if err == nil {
-			klog.Infof("Added MAC binding for %s on %s", gatewayIP, localnetGatewayNextHopPort)
-		} else {
-			klog.Errorf("Error in adding MAC binding for %s on %s: %v", gatewayIP, localnetGatewayNextHopPort, err)
+	for _, ifaddr := range gatewayIfAddrs {
+		err = initLocalGatewayNATRules(localnetGatewayNextHopPort, ifaddr.IP)
+		if err != nil {
+			return fmt.Errorf("failed to add NAT rules for localnet gateway (%v)", err)
 		}
 	}
 
-	err = initLocalGatewayNATRules(localnetGatewayNextHopPort, gatewayIP)
-	if err != nil {
-		return fmt.Errorf("failed to add NAT rules for localnet gateway (%v)", err)
-	}
-
 	if config.Gateway.NodeportEnable {
 		localAddrSet, err := getLocalAddrs()
 		if err != nil {
 			return err
 		}
 		err = n.watchLocalPorts(
-			newLocalPortWatcherData(gatewayIP, n.recorder, localAddrSet),
+			newLocalPortWatcherData(gatewayIfAddrs, n.recorder, localAddrSet),
```

Comment:
This is not an issue with this PR but it may be an issue with the earlier PR that refactored IP table rules into its  own file.  Ithe method  `newLocalPortWatcherData`  calls the method  `initLocalGatewayIPTables` which I think always installs IPv4 and IPv6 rules irrespective of which stack is enabled.  I think we may have to correct it to first check the stack(s) that are enabled before generating and installing IPT rules.